src/nvim/eval/encode.c

```diff
@@ -0,0 +1,1291 @@
+static const void *last_p_enc = NULL;
+
+/// Conversion setup for converting from last_p_enc to UTF-8
+static vimconv_T p_enc_conv = {
+  .vc_type = CONV_NONE,
+};
+
+/// Escape sequences used in JSON
+static const char escapes[][3] = {
+  [BS] = "\\b",
+  [TAB] = "\\t",
+  [NL] = "\\n",
+  [CAR] = "\\r",
+  ['"'] = "\\\"",
+  ['\\'] = "\\\\",
+  [FF] = "\\f",
+};
+
+static const char xdigits[] = "0123456789ABCDEF";
+
+/// Convert given string to JSON string
+///
+/// @param[out]  gap  Garray where result will be saved.
+/// @param[in]  buf  Converted string.
+/// @param[in]  len  Converted string length.
+///
+/// @return OK in case of success, FAIL otherwise.
+static inline int convert_to_json_string(garray_T *const gap,
+                                         const char *const buf,
+                                         const size_t len)
+  FUNC_ATTR_NONNULL_ARG(1) FUNC_ATTR_ALWAYS_INLINE
+{
+  const char *utf_buf = buf;
+  if (utf_buf == NULL) {
+    ga_concat(gap, "\"\"");
+  } else {
+    size_t utf_len = len;
+    char *tofree = NULL;
+    if (last_p_enc != (const void *) p_enc) {
+      p_enc_conv.vc_type = CONV_NONE;
+      convert_setup(&p_enc_conv, p_enc, "utf-8");
+      p_enc_conv.vc_fail = true;
+      last_p_enc = p_enc;
+    }
+    if (p_enc_conv.vc_type != CONV_NONE) {
+      tofree = string_convert(&p_enc_conv, buf, &utf_len);
+      if (tofree == NULL) {
+        emsgf(_("E474: Failed to convert string \"%.*s\" to UTF-8"),
+              utf_len, utf_buf);
+        return FAIL;
+      }
+      utf_buf = tofree;
+    }
+    size_t str_len = 0;
+    // Encode character as \u0000 if
+    // 1. It is an ASCII control character (0x0 .. 0x1F).
+    // 2. &encoding is not UTF-8 and code point is above 0x7F.
+    // 3. &encoding is UTF-8 and code point is printable according to
+    //    utf_printable().
+    // This is done to make it possible to :echo values when &encoding is not
+    // UTF-8.
+#define ENCODE_RAW(p_enc_conv, ch) \
+    (ch >= 0x20 && (p_enc_conv.vc_type == CONV_NONE \
+                    ? utf_printable(ch) \
+                    : ch < 0x7F))
+    for (size_t i = 0; i < utf_len;) {
+      const int ch = utf_ptr2char(utf_buf + i);
+      const size_t shift = (ch == 0? 1: utf_ptr2len(utf_buf + i));
+      assert(shift > 0);
+      i += shift;
+      switch (ch) {
+        case BS:
+        case TAB:
+        case NL:
+        case FF:
+        case CAR:
+        case '"':
+        case '\\': {
+          str_len += 2;
+          break;
+        }
+        default: {
+          if (ch > 0x7F && shift == 1) {
+            emsgf(_("E474: String \"%.*s\" contains byte that does not start "
+                    "any UTF-8 character"),
+                  utf_len - (i - shift), utf_buf + i - shift);
+            xfree(tofree);
+            return FAIL;
+          } else if ((SURROGATE_HI_START <= ch && ch <= SURROGATE_HI_END)
+                     || (SURROGATE_LO_START <= ch && ch <= SURROGATE_LO_END)) {
```

Comment:
Condition could be shortened to `SURROGATE_HI_START <= ch && ch <= SURROGATE_LO_END`